policyeval/policyserver.go

```diff
@@ -0,0 +1,141 @@
+		if time.Since(resp.Timestamp) < 5*time.Minute {
+			return &PolicyServerResponse{Recommendation: resp.Recommendation}, true
+		}
+		ps.cacheLock.Lock()
+		delete(ps.EventCache, evtID)
+		ps.cacheLock.Unlock()
+	}
+	return nil, false
+}
+
+func (ps *PolicyServer) cacheRecommendation(evtID id.EventID, entry *psCacheEntry) {
+	ps.cacheLock.Lock()
+	defer ps.cacheLock.Unlock()
+	ps.EventCache[evtID] = entry
+	if len(ps.EventCache) > 1000 {
+		// clear out old entries to save space
+		for k := range ps.EventCache {
+			if time.Since(ps.EventCache[k].Timestamp) > 5*time.Minute {
+				delete(ps.EventCache, k)
+			}
+		}
+	}
+}
+
+type PSRecommendation string
+
+const (
+	PSRecommendationOk   PSRecommendation = "ok"
+	PSRecommendationSpam PSRecommendation = "spam"
+)
+
+type PolicyServerResponse struct {
+	Recommendation PSRecommendation `json:"recommendation"`
+	policy         *policylist.Match
+}
+
+var (
+	respPSOk = &PolicyServerResponse{Recommendation: PSRecommendationOk}
+)
+
+// MSC4284: https://github.com/matrix-org/matrix-spec-proposals/pull/4284
+
+func (ps *PolicyServer) getRecommendation(ctx context.Context, evtID id.EventID, pdu *util.EventPDU, evaluator *PolicyEvaluator) *PolicyServerResponse {
+	logger := zerolog.Ctx(ctx).With().Stringer("room_id", pdu.RoomID).Stringer("event_id", evtID).Logger()
+	watchedLists := evaluator.GetWatchedLists()
+	match := evaluator.Store.MatchUser(watchedLists, pdu.Sender)
+	res := &PolicyServerResponse{Recommendation: PSRecommendationSpam}
+	if match != nil {
+		res.policy = &match
+		return res
+	}
+	match = evaluator.Store.MatchServer(watchedLists, pdu.Sender.Homeserver())
+	if match != nil {
+		res.policy = &match
+		return res
+	}
+	// In theory, if a room is taken down and uses the policy server, we can prevent them sending further events
+	match = evaluator.Store.MatchRoom(watchedLists, pdu.RoomID)
+	if match != nil {
+		res.policy = &match
+		return res
+	}
+	// todo, in future, check against protections?
+
+	// Event seems fine.
+	logger.Trace().Msg("event accepted")
+	return respPSOk
+}
+
+func (ps *PolicyServer) HandleCheck(ctx context.Context, evtID id.EventID, pdu *util.EventPDU, evaluator *PolicyEvaluator, redact bool) (res *PolicyServerResponse, err error) {
+	if r, ok := ps.getCachedRecommendation(evtID); ok {
+		return r, nil
+	}
+	logger := zerolog.Ctx(ctx).With().Stringer("room_id", pdu.RoomID).Stringer("event_id", evtID).Logger()
+	logger.Trace().Interface("event", pdu).Msg("received check for protected room")
+	res = ps.getRecommendation(ctx, evtID, pdu, evaluator)
+	ps.cacheRecommendation(evtID, &psCacheEntry{
+		Recommendation: res.Recommendation,
+		Timestamp:      time.Now(),
+		RoomID:         pdu.RoomID,
+		PDU:            pdu,
+	})
+	if res.Recommendation == PSRecommendationSpam {
+		logger.Warn().Stringer("recommendations", res.policy.Recommendations()).Msg("event rejected for spam")
+		if redact {
+			if _, err := evaluator.Bot.RedactEvent(ctx, pdu.RoomID, evtID); err != nil {
```

Comment:
Can this result in sending duplicate redaction events?

Reply:
No, the result is cached before this call, meaning subsequent *and* concurrent calls will get the cached version (by nature of the mutex only allowing one RW at a time) rather than re-running this flow.
If the event is checked again 5 minutes or more later, then a duplicate redact could be sent, but it's also plausible that if a server is checked the event again later on, that they missed the redaction too.
I haven't stress tested this due to the signature verification issue mentioned in another comment (and also lack of server-side adoption) so I'm only talking hypotheticals.

Reply:
`HandleCheck` itself doesn't seem to be inside a mutex, only the cache fetches are. If you have two calls in parallel, you'll get two `getCachedRecommendation` calls (both won't find anything), then 2 `getRecommendation`s, 2 `cacheRecommendation`s and finally 2 redactions

The pattern I'd probably go for is storing a blank item in the cache map immediately and having a lock inside the "cache entry" for the rest of HandleCheck